### PR TITLE
Login history feature

### PR DESCRIPTION
**Pull request by johnayinde**
completed the login history feature 

-user browser information such as OS name, time of making the request, browser name, IP address and country is sent along with the login credentials

-the info is saved in a list for every subsequent login requests
-provided endpoint to get the user login histories

### CODE REVIEW

File: src/common/interface/userInfo.interface.ts [hunk 1]
```diff
@@ -1,7 +1,14 @@
-export interface UserInfo {
+import { ApiProperty } from '@nestjs/swagger';
+
+export class UserInfo {
```
Comment (Jude-adesulu) [inline src/common/interface/userInfo.interface.ts:3]: Use class-validator on these properties, some like @IsString()

File: src/user/user.controller.ts [hunk 2]
```diff
@@ -1,31 +1,56 @@
...
+  Put,
+} from '@nestjs/common';
+import { ApiOkResponse, ApiOperation, ApiTags } from '@nestjs/swagger';
 import { ZuAppResponse } from '../common/helpers/response';
 import { UserDto } from '../user/dto/user.dto';
 import { Serialize } from '../interceptors/serialize.interceptor';
 import { UserService } from './user.service';
 import { UpdateUserDto } from './dto/update-user.dto';
-
+import { UserInfo } from './../common/interface/userInfo.interface';
```
Comment (Jude-adesulu) [inline src/user/user.controller.ts:18]: You are exporting a class right? Why is the name covention users.interface. you can use a dto as the extension name instead

File: src/user/user.controller.ts [hunk 3]
```diff
@@ -1,31 +1,56 @@
...
+      'User update successful',
+      '200',
+    );
+  }
 
+  @Get('history/:userId')
+  @ApiOkResponse({
+    type: UserInfo,
+    description: 'Get all histories',
+    isArray: true,
```
Comment (Jude-adesulu) [inline src/user/user.controller.ts:50]: Are you returning an array or an object? Cause the response in your service is an object type

**Review by Jude-adesulu — CHANGES_REQUESTED** [review]

File: src/user/user.service.ts [hunk 4]
```diff
@@ -68,4 +68,19 @@ export class UserService {
...
+        ZuAppResponse.NotFoundRequest(
+          'Internal server error',
+          'user not found',
+          '404',
+        ),
+      );
+    }
+
+    return { histories: user.history };
+  }
```
Comment (Jude-adesulu) [inline src/user/user.service.ts:85]: Use the ZuAppResponse custom response here

**Review by Jude-adesulu — CHANGES_REQUESTED** [review]

**Review by SunmadeBabafemi — APPROVED** [review]
I believe it is now left for the frontend to fetch those data each time a user logs in right?....

nice work

**Review by andyjrII — APPROVED** [review]

**Review by Jude-adesulu — APPROVED** [review]
Good job bro

File: src/user/user.controller.ts (hunk 2 above)
Comment (johnayinde) [reply src/user/user.controller.ts]: Changes made All fixed

**Comment by johnayinde** [comment]
> Good job bro

**Comment by johnayinde** [comment]
@PhilemonBrain

**Review by PhilemonBrain — APPROVED** [review]

**Pull request merged**